crates/noirc_frontend/src/parser/parser.rs

```diff
@@ -235,12 +236,22 @@ fn lambda_return_type() -> impl NoirParser<UnresolvedType> {
         .map(|ret| ret.unwrap_or(UnresolvedType::Unspecified))
 }
 
-fn function_return_type() -> impl NoirParser<(AbiVisibility, UnresolvedType)> {
+fn function_return_type() -> impl NoirParser<(AbiVisibility, AbiDistinctness, UnresolvedType)> {
     just(Token::Arrow)
         .ignore_then(optional_visibility())
+        .then(optional_distinctness())
         .then(parse_type())
         .or_not()
-        .map(|ret| ret.unwrap_or((AbiVisibility::Private, UnresolvedType::Unit)))
+        .map(|ret| {
+            ret.map(|((visibility, distinctness), return_type)| {
+                (visibility, distinctness, return_type)
+            })
```

Comment:
You can optionally remove this map if you want to save some code. This is only used in function_definition which is already a mess of nested tuples anyways.